kips/kip-2.md

```diff
@@ -0,0 +1,462 @@
+
+_Note: I just copied the ACL code from an existing project. There could be a simpler solution for admin management if this seems too complex._
+
+```solidity
+/**
+ * Interface for ACL.
+ */
+interface IACL {
+  // admin
+  function isAdmin(address _addr) external view returns (bool);
+  function proposeNewAdmin(address _addr) external;
+  function cancelNewAdminProposal(address _addr) external;
+  function acceptAdminRole() external;
+  function removeAdmin(address _addr) external;
+}
+
+/**
+ * Access-control list
+ */
+contract ACL is IACL {
+  mapping (address => bool) public admins;
+  mapping (address => bool) public pendingAdmins;
+  uint256 public numAdmins;
+
+  event AdminProposed(address indexed addr);
+  event AdminProposalCancelled(address indexed addr);
+  event AdminProposalAccepted(address indexed addr);
+  event AdminRemoved(address indexed addr);
+
+  modifier assertIsAdmin () {
+    require(admins[msg.sender], 'unauthorized - must be admin');
+    _;
+  }
+
+  constructor () public {
+    admins[msg.sender] = true;
+    numAdmins = 1;
+  }
+
+  function isAdmin(address _addr) public view returns (bool) {
+    return admins[_addr];
+  }
+
+  function proposeNewAdmin(address _addr) public assertIsAdmin {
+    require(!admins[_addr], 'already an admin');
+    require(!pendingAdmins[_addr], 'already proposed as an admin');
+    pendingAdmins[_addr] = true;
+    emit AdminProposed(_addr);
+  }
+
+  function cancelNewAdminProposal(address _addr) public assertIsAdmin {
+    require(pendingAdmins[_addr], 'not proposed as an admin');
+    pendingAdmins[_addr] = false;
+    emit AdminProposalCancelled(_addr);
+  }
+
+  function acceptAdminRole() public {
+    require(pendingAdmins[msg.sender], 'not proposed as an admin');
+    pendingAdmins[msg.sender] = false;
+    admins[msg.sender] = true;
+    numAdmins++;
+    emit AdminProposalAccepted(msg.sender);
+  }
+
+  function removeAdmin(address _addr) public assertIsAdmin {
+    require(1 < numAdmins, 'cannot remove last admin');
+    require(_addr != msg.sender, 'cannot remove oneself');
+    require(admins[_addr], 'not an admin');
+    admins[_addr] = false;
+    numAdmins--;
+    emit AdminRemoved(_addr);
+  }
+}
+
+/**
+ * Access control helper base class.
+ */
+contract AccessControl is EternalStorage {
+  modifier assertIsAdmin () {
+    require(acl().isAdmin(msg.sender), 'must be admin');
+    _;
+  }
+
+  constructor (address _acl) public {
+    dataAddress["acl"] = _acl;
+  }
+
+  function acl () internal view returns (IACL) {
+    return IACL(dataAddress["acl"]);
+  }
+}
+```
+
+**UserPot, IUserPot, UserPotImpl**
+
+The user pot.
+
+Note that `UserPot` doesn't implement `IUserPot` itself since it's going to be a proxy contract which delegates all calls to a `UserPotImpl` instance. Also note that `upgrade()` can only be called by an administrator.
+
```

Comment:
Is it possible for an admin to steal all user's pot?

Reply:
Yes, by upgrading the contract to something which allows that. Only way to mitigate this would be store the money separately in yet another contract. Can reduce the risk by requiring multiple admin signatures in order to perform an upgrade.

Reply:
Chris suggested that we could also put upgrades on a time delay so that users would always have time to be alerted and withdraw in case of a malicious upgrade. We could even combine some sort of "Emergency shutdown" that admins can cancel all the ongoing event and empty out all pots by sending back to everyone. This could work in the worst case scenario where we have to completely abandon migrate off of current platform to the new one.